galsim/phase_screens.py

```diff
@@ -0,0 +1,652 @@
+        if 'direction' not in kwargs:
+            kwargs['direction'] = [0*galsim.degrees]*len(kwargs['speed'])
+        kwargs['vx'], kwargs['vy'] = zip(*[v*d.sincos()
+                                           for v, d in zip(kwargs['speed'],
+                                                           kwargs['direction'])])
+        del kwargs['speed']
+        del kwargs['direction']
+
+    # Determine broadcast size
+    nmax = max(len(v) for v in kwargs.values() if hasattr(v, '__len__'))
+
+    # Broadcast r0_500 here, since logical combination of indiv layers' r0s is complex:
+    if len(kwargs['r0_500']) == 1:
+        kwargs['r0_500'] = [nmax**(3./5) * kwargs['r0_500'][0]] * nmax
+
+    if rng is None:
+        rng = galsim.BaseDeviate()
+    kwargs['rng'] = [galsim.BaseDeviate(rng.raw()) for i in xrange(nmax)]
+    return galsim.PhaseScreenList(AtmosphericScreen(**kw) for kw in _lod_to_dol(kwargs, nmax))
+
+
+# Some utilities for working with Zernike polynomials
+# Combinations.  n choose r.
+# See http://stackoverflow.com/questions/3025162/statistics-combinations-in-python
+# This is J. F. Sebastian's answer.
+def _nCr(n, r):
+    if 0 <= r <= n:
+        ntok = 1
+        rtok = 1
+        for t in xrange(1, min(r, n - r) + 1):
+            ntok *= n
+            rtok *= t
+            n -= 1
+        return ntok // rtok
+    else:
+        return 0
+
+
+# This function stolen from https://github.com/tvwerkhoven/libtim-py/blob/master/libtim/zern.py
+def _noll_to_zern(j):
+    """
+    Convert linear Noll index to tuple of Zernike indices.
+    j is the linear Noll coordinate, n is the radial Zernike index and m is the azimuthal Zernike
+    index.
+    @param [in] j Zernike mode Noll index
+    @return (n, m) tuple of Zernike indices
+    @see <https://oeis.org/A176988>.
+    """
+    if (j == 0):
+        raise ValueError("Noll indices start at 1. 0 is invalid.")
+
+    n = 0
+    j1 = j-1
+    while (j1 > n):
+        n += 1
+        j1 -= n
+    m = (-1)**j * ((n % 2) + 2 * int((j1+((n+1) % 2)) / 2.0))
+    return (n, m)
+
+
+def _zern_norm(n, m):
+    """Normalization coefficient for zernike (n, m).
+
+    Defined such that \int_{unit disc} Z(n1, m1) Z(n2, m2) dA = \pi if n1==n2 and m1==m2 else 0.0
+    """
+    if m == 0:
+        return np.sqrt(1./(n+1))
+    else:
+        return np.sqrt(1./(2.*n+2))
+
+
+def _zern_rho_coefs(n, m):
+    """Compute coefficients of radial part of Zernike (n, m).
+    """
+    kmax = (n-abs(m))/2
+    A = [0]*(n+1)
+    val = _nCr(n,kmax) # The value for k = 0 in the equation below.
+    norm = _zern_norm(n,m)
+    for k in xrange(kmax):
+        # val = (-1)**k * _nCr(n-k, k) * _nCr(n-2*k, kmax-k) / _zern_norm(n, m)
+        # The above formula is faster as a recurrence relation:
+        A[n-2*k] = val / norm
+        # Don't use *= since the factor is not an integer, but the result is.
+        val = -val * (kmax-k)*(n-kmax-k) // ((n-k)*(k+1))
+    A[n-2*kmax] = val / norm
+    return A
```

Comment:
Even though this function is never anywhere close to a tall pole for optimization, I felt compelled to turn it into a more efficient recurrence relation rather than calculating nCr repeatedly.